cmake/compiler/gcc.cmake

```diff
@@ -50,56 +58,76 @@ foreach(file_name include include-fixed)
+  include(${ZEPHYR_BASE}/cmake/gcc-m-cpu.cmake)
 
-if("${ARCH}" STREQUAL "arm")
-  list(APPEND TOOLCHAIN_C_FLAGS
-    -mthumb
-    -mcpu=${GCC_M_CPU}
-    )
+  if("${ARCH}" STREQUAL "arm")
+    list(APPEND TOOLCHAIN_C_FLAGS
+      -mthumb
+      -mcpu=${GCC_M_CPU}
+      )
 
-  include(${ZEPHYR_BASE}/cmake/fpu-for-gcc-m-cpu.cmake)
+    include(${ZEPHYR_BASE}/cmake/fpu-for-gcc-m-cpu.cmake)
 
-  if(CONFIG_FLOAT)
-    list(APPEND TOOLCHAIN_C_FLAGS -mfpu=${FPU_FOR_${GCC_M_CPU}})
-    if    (CONFIG_FP_SOFTABI)
-      list(APPEND TOOLCHAIN_C_FLAGS -mfloat-abi=soft)
-    elseif(CONFIG_FP_HARDABI)
-      list(APPEND TOOLCHAIN_C_FLAGS -mfloat-abi=hard)
+    if(CONFIG_FLOAT)
+      list(APPEND TOOLCHAIN_C_FLAGS -mfpu=${FPU_FOR_${GCC_M_CPU}})
+      if    (CONFIG_FP_SOFTABI)
+        list(APPEND TOOLCHAIN_C_FLAGS -mfloat-abi=soft)
+      elseif(CONFIG_FP_HARDABI)
+        list(APPEND TOOLCHAIN_C_FLAGS -mfloat-abi=hard)
+      endif()
     endif()
+  elseif("${ARCH}" STREQUAL "arc")
+    list(APPEND TOOLCHAIN_C_FLAGS
+      -mcpu=${GCC_M_CPU}
+      )
   endif()
-elseif("${ARCH}" STREQUAL "arc")
-  list(APPEND TOOLCHAIN_C_FLAGS
-	  -mcpu=${GCC_M_CPU}
+
+  execute_process(
+    COMMAND ${CMAKE_C_COMPILER} ${TOOLCHAIN_C_FLAGS} --print-libgcc-file-name
+    OUTPUT_VARIABLE LIBGCC_FILE_NAME
+    OUTPUT_STRIP_TRAILING_WHITESPACE
     )
-endif()
 
-execute_process(
-  COMMAND ${CMAKE_C_COMPILER} ${TOOLCHAIN_C_FLAGS} --print-libgcc-file-name
-  OUTPUT_VARIABLE LIBGCC_FILE_NAME
-  OUTPUT_STRIP_TRAILING_WHITESPACE
-  )
+  assert_exists(LIBGCC_FILE_NAME)
+
+  get_filename_component(LIBGCC_DIR ${LIBGCC_FILE_NAME} DIRECTORY)
 
-execute_process(
-  COMMAND ${CMAKE_C_COMPILER} ${TOOLCHAIN_C_FLAGS} --print-multi-directory
-  OUTPUT_VARIABLE NEWLIB_DIR
-  OUTPUT_STRIP_TRAILING_WHITESPACE
-  )
+  assert_exists(LIBGCC_DIR)
 
-assert_exists(LIBGCC_FILE_NAME)
+  LIST(APPEND LIB_INCLUDE_DIR "-L\"${LIBGCC_DIR}\"")
+  LIST(APPEND TOOLCHAIN_LIBS gcc)
 
-get_filename_component(LIBGCC_DIR ${LIBGCC_FILE_NAME} DIRECTORY)
+  if(SYSROOT_DIR)
+    # The toolchain has specified a sysroot dir that we can use to set
+    # the libc path's
+    execute_process(
+      COMMAND ${CMAKE_C_COMPILER} ${TOOLCHAIN_C_FLAGS} --print-multi-directory
+      OUTPUT_VARIABLE NEWLIB_DIR
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+      )
 
-assert_exists(LIBGCC_DIR)
+    set(LIBC_LIBRARY_DIR "\"${SYSROOT_DIR}\"/lib/${NEWLIB_DIR}")
+    set(LIBC_INCLUDE_DIR ${SYSROOT_DIR}/include)
+  endif()
+
+elseif("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "xcc")
+  # Include XCC standard libraries so that users used to Xplorer IDE can port
+  # their code easily
+  set(TOOLCHAIN_INCLUDES
+      ${TOOLCHAIN_HOME}/lib/xcc/include
+      ${TOOLCHAIN_HOME}/xtensa-elf/include
+      ${XCC_BUILD}/xtensa-elf/arch/include
+      ${XCC_BUILD}/xtensa-elf/include
+      )
 
-LIST(APPEND LIB_INCLUDE_DIR "-L\"${LIBGCC_DIR}\"")
-LIST(APPEND TOOLCHAIN_LIBS gcc)
+  LIST(APPEND TOOLCHAIN_LIBS gcc)
+  LIST(APPEND TOOLCHAIN_LIBS hal)
```

Comment:
HAL libraries, their header file include paths, and their lib include directories should not be added by the toolchain.

If they are added by the toolchain it will not be possible to opt-out of using the hal.

Instead, they should be added through Kconfig.

This is an example of the same situation that xcc's 'hal' is in:

https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/qmsi/CMakeLists.txt#L12

It is a toolchain vendor-supplied binary 'hal' library with an include directory for header files and an include directory for the binary lib itself. This library must be organized in the same manner if possible.

Reply:
I had a relook at what i am doing for XCC in this file and i see that only below 2 lines are required and i can omit everything else.
  LIST(APPEND TOOLCHAIN_LIBS gcc)
  LIST(APPEND TOOLCHAIN_LIBS hal)

The other things included seem to be for facilitating the users used to Xplorer IDE (XCC standard libraries) to port their code easily.

Reply:
Had a look at the Kconfig thing. I think it may not be applicable in this case as there is nothing configurable here. If you plan to use XCC for your build, you just get a set of preconfigured things. If you want to use GCC, then apparently this HAL is not required.